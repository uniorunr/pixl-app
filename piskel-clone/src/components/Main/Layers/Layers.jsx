import React, { Component } from 'react';
import './Layers.scss';
import PropTypes from 'prop-types';

const saveLayerData = (keys, index, framesData, frames) => {
  const currLayerFramesData = framesData;
  const layerKey = `layer${keys[index]}`;
  currLayerFramesData[layerKey] = frames.map(item => item.toDataURL());
  sessionStorage.setItem('framesData', JSON.stringify(framesData));
};

const restoreFrames = (frames, keys, active, framesData) => {
  const canvas = document.querySelector('#canvas');
  const canvasContext = canvas.getContext('2d');
  canvasContext.imageSmoothingEnabled = false;
  canvas.getContext('2d').clearRect(0, 0, canvas.width, canvas.height);
  frames.forEach((frame, index) => {
    const frameContext = frame.getContext('2d');
    frameContext.imageSmoothingEnabled = false;
    frameContext.clearRect(0, 0, frame.width, frame.height);
    const img = new Image();
    const layerId = `layer${keys[active]}`;
    if (!framesData[layerId]) return;
    const URI = framesData[layerId][index];
    if (URI) {
      img.src = URI;
      img.onload = () => {
        frameContext.drawImage(img, 0, 0);
        saveLayerData(keys, active, framesData, frames);
        if (frame.classList.contains('frame__canvas_active')) {
          canvasContext.drawImage(img, 0, 0, canvas.width, canvas.height);
        }
      };
    }
  });
};

class Layers extends Component {
  state = {
    layers: [],
    active: +sessionStorage.getItem('activeLayer') || 0,
    keys: JSON.parse(sessionStorage.getItem('layerKeys')) || [0],
  };

  componentDidMount = () => {
    const { active, keys } = this.state;
    const { framesData, framesArray, updateLayerKeys } = this.props;
    const initLayer = (
      <button
        className="layers-section__button layers-section__button_active"
        type="button"
        key="0"
        id="layer0"
        onClick={this.makeActive}
      >
        Layer 1
      </button>
    );
    this.setState({
      layers: [initLayer],
    });
    updateLayerKeys([0]);
    if (!sessionStorage.getItem('framesData')) {
      saveLayerData(keys, active, framesData, framesArray);
    } else {
      restoreFrames(framesArray, keys, active, framesData);
      saveLayerData(keys, active, framesData, framesArray);
    }
    if (!sessionStorage.getItem('layerKeys')) {
      sessionStorage.setItem('layerKeys', JSON.stringify([0]));
    }
  };

  componentDidUpdate() {
    const { keys, active } = this.state;
    const { framesData, framesArray } = this.props;
    saveLayerData(keys, active, framesData, framesArray);
  }

  handleAddButton = () => {
    const { layers, keys, active } = this.state;
    const {
      framesData,
      framesArray,
      updateLayerKeys,
      updateActiveLayer,
    } = this.props;
    saveLayerData(keys, active, framesData, framesArray);
    const canvas = document.querySelector('#canvas');
    canvas.getContext('2d').clearRect(0, 0, canvas.width, canvas.height);
    framesArray.forEach((frame) => {
      frame.getContext('2d').clearRect(0, 0, frame.width, frame.height);
    });
    const newLayer = (
      <button
        className="layers-section__button"
        type="button"
        key={Math.max(...keys) + 1}
        id={`layer${Math.max(...keys) + 1}`}
        data-id={Math.max(...keys) + 1}
        onClick={this.makeActive}
      >
        Layer {Math.max(...keys) + 1}
      </button>
    );
    const updatedKeys = [...keys, Math.max(...keys) + 1];
    updateLayerKeys(updatedKeys);
    updateActiveLayer(keys.length);
    this.setState({
      layers: [...layers, newLayer],
      keys: updatedKeys,
      active: keys.length,
    });
    sessionStorage.setItem('layerKeys', JSON.stringify(updatedKeys));
    sessionStorage.setItem('activeLayer', `${keys.length}`);
  };

  handleRemoveButton = () => {
    const {
      framesData,
      framesArray,
      updateLayerKeys,
      updateActiveLayer,
    } = this.props;
    const { layers, keys, active } = this.state;
    delete framesData[`layer${keys[active]}`];
    if (keys.length > 1) {
      const newActive = keys.length - 2;
      layers.splice(active, 1);
      keys.splice(active, 1);
      restoreFrames(framesArray, keys, newActive, framesData);
      this.setState({
        layers: [...layers],
        keys: [...keys],
        active: newActive,
      });
      updateLayerKeys([...keys]);
      updateActiveLayer(newActive);
      sessionStorage.setItem('layerKeys', JSON.stringify([...keys]));
      sessionStorage.setItem('activeLayer', `${newActive}`);
    }
  };

  handleMoveUp = () => {
    const { layers, keys, active } = this.state;
    const { updateLayerKeys, updateActiveLayer } = this.props;
    const index = active;
    if (index > 0) {
      [layers[index - 1], layers[index]] = [layers[index], layers[index - 1]];
      [keys[index - 1], keys[index]] = [keys[index], keys[index - 1]];
      this.setState({
        layers: [...layers],
        active: active - 1,
        keys: [...keys],
      });
      updateLayerKeys([...keys]);
      updateActiveLayer(active - 1);
      sessionStorage.setItem('layerKeys', JSON.stringify([...keys]));
      sessionStorage.setItem('activeLayer', `${active - 1}`);
    }
  };

  handleMoveDown = () => {
    const { layers, keys, active } = this.state;
    const { updateLayerKeys, updateActiveLayer } = this.props;
    const index = active;
    if (index < keys.length - 1) {
      [layers[index + 1], layers[index]] = [layers[index], layers[index + 1]];
      [keys[index + 1], keys[index]] = [keys[index], keys[index + 1]];
      this.setState({
        layers: [...layers],
        active: active + 1,
        keys: [...keys],
      });
      updateLayerKeys([...keys]);
      updateActiveLayer(active + 1);
      sessionStorage.setItem('layerKeys', JSON.stringify([...keys]));
      sessionStorage.setItem('activeLayer', `${active + 1}`);
    }
  };

  handleMergeButton = () => {
    const { framesData, framesArray, updateLayerKeys } = this.props;
    const { layers, keys, active } = this.state;
    const canvas = document.querySelector('#canvas');
    saveLayerData(keys, active, framesData, framesArray);
    if (keys[active + 1] >= 0) {
      canvas.getContext('2d').clearRect(0, 0, canvas.width, canvas.height);
      framesArray.forEach((frame, index) => {
        const frameContext = frame.getContext('2d');
        frameContext.imageSmoothingEnabled = false;
        frameContext.clearRect(0, 0, frame.width, frame.height);
        const frontImg = new Image();
        const backImg = new Image();
        const frontLayerId = `layer${keys[active]}`;
        const backLayerId = `layer${keys[active + 1]}`;
        const frontURI = framesData[frontLayerId][index];
        const backURI = framesData[backLayerId][index];
        if (backURI) {
          backImg.src = backURI;
          backImg.onload = () => {
            frameContext.drawImage(backImg, 0, 0);
            saveLayerData(keys, active, framesData, framesArray);
            if (frame.classList.contains('frame__canvas_active')) {
              const canvasContext = canvas.getContext('2d');
              canvasContext.imageSmoothingEnabled = false;
              canvasContext.drawImage(
                backImg,
                0,
                0,
                canvas.width,
                canvas.height,
              );
            }
            if (frontURI) {
              frontImg.src = frontURI;
              frontImg.onload = () => {
                frameContext.drawImage(frontImg, 0, 0);
                saveLayerData(keys, active, framesData, framesArray);
                if (frame.classList.contains('frame__canvas_active')) {
                  const canvasContext = canvas.getContext('2d');
                  canvasContext.imageSmoothingEnabled = false;
                  canvasContext.drawImage(
                    frontImg,
                    0,
                    0,
                    canvas.width,
                    canvas.height,
                  );
                }
              };
            }
          };
        }
      });
      delete framesData[`layer${keys[active + 1]}`];
      layers.splice(active + 1, 1);
      keys.splice(active + 1, 1);
      this.setState({
        layers: [...layers],
        keys: [...keys],
      });
      updateLayerKeys([...keys]);
      sessionStorage.setItem('layerKeys', JSON.stringify([...keys]));
    }
  };

  makeActive = ({ target }) => {
    const { framesData, framesArray, updateActiveLayer } = this.props;
    const { active, keys } = this.state;
    saveLayerData(keys, active, framesData, framesArray);
    const newActive = keys.indexOf(+target.dataset.id);
    if (keys.length > 1) {
      restoreFrames(framesArray, keys, newActive, framesData);
      saveLayerData(keys, newActive, framesData, framesArray);
    }
    this.setState({
      active: newActive,
    });
    updateActiveLayer(newActive);
    sessionStorage.setItem('activeLayer', `${newActive}`);
  };

  render() {
    const { active, keys } = this.state;

    return (
      <section className="layers-section">
        <div className="layers-section__controls">
          <button
            type="button"
            className="layers-section__add-button"
            onClick={this.handleAddButton}
          >
            <i className="fas fa-plus" />
          </button>
          <button
            type="button"
            className="layers-section__move-up"
            onClick={this.handleMoveUp}
          >
            <i className="fas fa-arrow-up" />
          </button>
          <button
            type="button"
            className="layers-section__move-down"
            onClick={this.handleMoveDown}
          >
            <i className="fas fa-arrow-down" />
          </button>
          <button
            type="button"
            className="layers-section__remove-button"
            onClick={this.handleRemoveButton}
          >
            <i className="fas fa-trash-alt" />
          </button>
        </div>
        <button
          type="button"
          className="layers-section__merge-button"
          onClick={this.handleMergeButton}
        >
          merge
        </button>
        <div className="layers-section__wrapper">
          {keys.map((item, index) => (
            <button
              className={`layers-section__button ${
                index === active ? 'layers-section__button_active' : ''
              }`}
              id={`layer${keys[index]}`}
              data-id={keys[index]}
              type="button"
              key={keys[index]}
              onClick={this.makeActive}
            >
              Layer {keys[index] + 1}
            </button>
          ))}
        </div>
      </section>
    );
  }
}

Layers.propTypes = {
  updateLayerKeys: PropTypes.func.isRequired,
  updateActiveLayer: PropTypes.func.isRequired,
  framesData: PropTypes.instanceOf(Object).isRequired,
  framesArray: PropTypes.instanceOf(Array).isRequired,
};

export default Layers;
