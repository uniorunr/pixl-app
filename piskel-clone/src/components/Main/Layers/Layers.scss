.layers-section {
  display: flex;
  justify-content: flex-start;
  flex-direction: column;
  align-items: center;
  margin: 20px;
  padding: 5px;
  background-color: white;
  border: 2px solid black;
  max-height: 50%;

  .layers-section__wrapper {
    width: 100%;
    display: flex;
    flex-direction: column;
    align-items: center;
    margin: 10px 0;
    padding: 0 5px;
    overflow-y: auto;

    .layers-section__button {
      width: 100%;
      margin: 5px;
      padding: 5px 10px;
      border: 2px solid black;
      background-color: white;
      cursor: pointer;
      font-family: "Roboto", "Arial", sans-serif;
      font-weight: 700;
      font-size: 1.4rem;
      text-align: left;
      outline: none;
    }

    .layers-section__button_active {
      border: 3px dashed black;
    }
  }

  .layers-section__controls {
    width: 100%;
    display: flex;
    justify-content: space-around;
    margin: 5px 0;

    .layers-section__add-button,
    .layers-section__remove-button,
    .layers-section__move-up,
    .layers-section__move-down {
      position: relative;
      top: 0;
      box-shadow: 2px 2px 0 black;
      border: 2px solid black;
      background: white;
      outline: none;
      cursor: pointer;
      font-size: 1.4rem;
      padding: 2px 5px;
    }

    .layers-section__add-button:active,
    .layers-section__remove-button:active,
    .layers-section__move-up:active,
    .layers-section__move-down:active {
      top: 2px;
      box-shadow: 0 0 0 black;
    }
  }
}
