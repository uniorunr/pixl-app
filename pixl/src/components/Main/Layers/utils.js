const saveLayerData = (keys, index, framesData, frames, updFunc) => {
  const layerKey = `layer${keys[index]}`;
  if (keys[index] !== undefined) {
    const tempFramesData = { ...framesData };
    tempFramesData[layerKey] = frames.map(item => item.toDataURL());
    updFunc(tempFramesData);
    sessionStorage.setItem('framesData', JSON.stringify(tempFramesData));
  }
};

const restoreFrames = (frames, keys, active, framesData, updFunc) => {
  const canvas = document.querySelector('#canvas');
  const canvasContext = canvas.getContext('2d');
  canvasContext.imageSmoothingEnabled = false;
  canvas.getContext('2d').clearRect(0, 0, canvas.width, canvas.height);
  frames.forEach((frame, index) => {
    const frameContext = frame.getContext('2d');
    frameContext.imageSmoothingEnabled = false;
    frameContext.clearRect(0, 0, frame.width, frame.height);
    const img = new Image();
    const layerId = `layer${keys[active]}`;
    if (!framesData[layerId]) return;
    const URI = framesData[layerId][index];
    if (URI) {
      img.src = URI;
      img.onload = () => {
        frameContext.drawImage(img, 0, 0);
        saveLayerData(keys, active, framesData, frames, updFunc);
        if (frame.classList.contains('frame__canvas_active')) {
          canvasContext.drawImage(img, 0, 0, canvas.width, canvas.height);
        }
      };
    }
  });
};

export { saveLayerData, restoreFrames };
