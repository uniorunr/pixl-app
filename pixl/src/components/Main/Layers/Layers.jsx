import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import * as actions from '../../../actions/actions';
import './Layers.scss';
import { saveLayerData, restoreFrames } from './utils';

class Layers extends Component {
  componentDidMount = () => {
    const {
      framesData,
      framesArray,
      layerKeys,
      activeLayer,
      updateFramesData,
    } = this.props;
    if (!sessionStorage.getItem('framesData')) {
      saveLayerData(
        layerKeys,
        activeLayer,
        framesData,
        framesArray,
        updateFramesData,
      );
    } else {
      restoreFrames(
        framesArray,
        layerKeys,
        activeLayer,
        framesData,
        updateFramesData,
      );
      saveLayerData(
        layerKeys,
        activeLayer,
        framesData,
        framesArray,
        updateFramesData,
      );
    }
    document.addEventListener('keydown', this.handleKeyboard);
  };

  componentWillUnmount() {
    document.removeEventListener('keydown', this.handleKeyboard);
  }

  handleKeyboard = ({ code, altKey }) => {
    const { layersShortcuts } = this.props;
    if (altKey) {
      if (code === layersShortcuts.add.shortcut) {
        this.handleAddButton();
      } else if (code === layersShortcuts.moveUp.shortcut) {
        this.handleMoveUp();
      } else if (code === layersShortcuts.moveDown.shortcut) {
        this.handleMoveDown();
      } else if (code === layersShortcuts.remove.shortcut) {
        this.handleRemoveButton();
      } else if (code === layersShortcuts.merge.shortcut) {
        this.handleMergeButton();
      }
    }
  };

  handleAddButton = () => {
    const {
      framesData,
      framesArray,
      updateLayerKeys,
      updateActiveLayer,
      canvas,
      layerKeys,
      activeLayer,
      updateFramesData,
    } = this.props;
    saveLayerData(
      layerKeys,
      activeLayer,
      framesData,
      framesArray,
      updateFramesData,
    );
    canvas.getContext('2d').clearRect(0, 0, canvas.width, canvas.height);
    framesArray.forEach((frame) => {
      frame.getContext('2d').clearRect(0, 0, frame.width, frame.height);
    });
    const updatedKeys = [...layerKeys, Math.max(...layerKeys) + 1];
    updateLayerKeys(updatedKeys);
    updateActiveLayer(layerKeys.length);
    saveLayerData(
      updatedKeys,
      layerKeys.length,
      framesData,
      framesArray,
      updateFramesData,
    );
    sessionStorage.setItem('layerKeys', JSON.stringify(updatedKeys));
    sessionStorage.setItem('activeLayer', `${layerKeys.length}`);
  };

  handleRemoveButton = () => {
    const {
      framesData,
      framesArray,
      updateLayerKeys,
      updateActiveLayer,
      layerKeys,
      activeLayer,
      updateFramesData,
    } = this.props;
    const tempFramesData = { ...framesData };
    const tempLayerKeys = [...layerKeys];
    delete tempFramesData[`layer${layerKeys[activeLayer]}`];
    updateFramesData(tempFramesData);
    if (layerKeys.length > 1) {
      const newActive = layerKeys.length - 2;
      tempLayerKeys.splice(activeLayer, 1);
      restoreFrames(
        framesArray,
        tempLayerKeys,
        newActive,
        tempFramesData,
        updateFramesData,
      );
      updateLayerKeys(tempLayerKeys);
      updateActiveLayer(newActive);
      sessionStorage.setItem('layerKeys', JSON.stringify(tempLayerKeys));
      sessionStorage.setItem('activeLayer', `${newActive}`);
    }
  };

  handleMoveUp = () => {
    const {
      updateLayerKeys,
      updateActiveLayer,
      layerKeys,
      activeLayer,
    } = this.props;
    const index = activeLayer;
    if (index > 0) {
      [layerKeys[index - 1], layerKeys[index]] = [
        layerKeys[index],
        layerKeys[index - 1],
      ];
      updateLayerKeys([...layerKeys]);
      updateActiveLayer(activeLayer - 1);
      sessionStorage.setItem('layerKeys', JSON.stringify([...layerKeys]));
      sessionStorage.setItem('activeLayer', `${activeLayer - 1}`);
    }
  };

  handleMoveDown = () => {
    const {
      updateLayerKeys,
      updateActiveLayer,
      layerKeys,
      activeLayer,
    } = this.props;
    const index = activeLayer;
    if (index < layerKeys.length - 1) {
      [layerKeys[index + 1], layerKeys[index]] = [
        layerKeys[index],
        layerKeys[index + 1],
      ];
      updateLayerKeys([...layerKeys]);
      updateActiveLayer(activeLayer + 1);
      sessionStorage.setItem('layerKeys', JSON.stringify([...layerKeys]));
      sessionStorage.setItem('activeLayer', `${activeLayer + 1}`);
    }
  };

  handleMergeButton = () => {
    const {
      framesData,
      framesArray,
      updateLayerKeys,
      updateActiveLayer,
      canvas,
      layerKeys,
      activeLayer,
      updateFramesData,
    } = this.props;
    saveLayerData(
      layerKeys,
      activeLayer,
      framesData,
      framesArray,
      updateFramesData,
    );
    if (layerKeys[activeLayer + 1] >= 0) {
      canvas.getContext('2d').clearRect(0, 0, canvas.width, canvas.height);
      framesArray.forEach((frame, index) => {
        const frameContext = frame.getContext('2d');
        frameContext.imageSmoothingEnabled = false;
        frameContext.clearRect(0, 0, frame.width, frame.height);
        const frontImg = new Image();
        const backImg = new Image();
        const frontLayerId = `layer${layerKeys[activeLayer]}`;
        const backLayerId = `layer${layerKeys[activeLayer + 1]}`;
        const frontURI = framesData[frontLayerId][index];
        const backURI = framesData[backLayerId][index];
        const tempFramesData = { ...framesData };
        const tempLayerKeys = [...layerKeys];
        delete tempFramesData[`layer${layerKeys[activeLayer + 1]}`];
        updateFramesData(tempFramesData);
        tempLayerKeys.splice(activeLayer + 1, 1);
        updateLayerKeys(tempLayerKeys);
        updateActiveLayer(activeLayer);
        if (backURI) {
          backImg.src = backURI;
          backImg.onload = () => {
            frameContext.drawImage(backImg, 0, 0);
            saveLayerData(
              tempLayerKeys,
              activeLayer,
              tempFramesData,
              framesArray,
              updateFramesData,
            );
            if (frame.classList.contains('frame__canvas_active')) {
              const canvasContext = canvas.getContext('2d');
              canvasContext.imageSmoothingEnabled = false;
              canvasContext.drawImage(
                backImg,
                0,
                0,
                canvas.width,
                canvas.height,
              );
            }
            if (frontURI) {
              frontImg.src = frontURI;
              frontImg.onload = () => {
                frameContext.drawImage(frontImg, 0, 0);
                saveLayerData(
                  tempLayerKeys,
                  activeLayer,
                  tempFramesData,
                  framesArray,
                  updateFramesData,
                );
                if (frame.classList.contains('frame__canvas_active')) {
                  const canvasContext = canvas.getContext('2d');
                  canvasContext.imageSmoothingEnabled = false;
                  canvasContext.drawImage(
                    frontImg,
                    0,
                    0,
                    canvas.width,
                    canvas.height,
                  );
                }
              };
            }
          };
        }
      });
      const tempFramesData = { ...framesData };
      const tempLayerKeys = [...layerKeys];
      delete tempFramesData[`layer${layerKeys[activeLayer + 1]}`];
      updateFramesData(tempFramesData);
      tempLayerKeys.splice(activeLayer + 1, 1);
      updateLayerKeys(tempLayerKeys);
      updateActiveLayer(activeLayer);
      sessionStorage.setItem('framesData', JSON.stringify(tempFramesData));
      sessionStorage.setItem('layerKeys', JSON.stringify(tempLayerKeys));
      sessionStorage.setItem('activeLayer', `${activeLayer}`);
    }
  };

  makeActive = ({ target }) => {
    const {
      framesData,
      framesArray,
      updateLayerKeys,
      updateActiveLayer,
      layerKeys,
      activeLayer,
      updateFramesData,
    } = this.props;
    saveLayerData(
      layerKeys,
      activeLayer,
      framesData,
      framesArray,
      updateFramesData,
    );
    const newActive = layerKeys.indexOf(+target.dataset.id);
    if (layerKeys.length > 1) {
      restoreFrames(
        framesArray,
        layerKeys,
        newActive,
        framesData,
        updateFramesData,
      );
      saveLayerData(
        layerKeys,
        newActive,
        framesData,
        framesArray,
        updateFramesData,
      );
    }
    updateLayerKeys([...layerKeys]);
    updateActiveLayer(newActive);
    sessionStorage.setItem('activeLayer', `${newActive}`);
  };

  render() {
    const { layerKeys, activeLayer } = this.props;

    return (
      <section className="layers-section">
        <div className="layers-section__controls">
          <button
            type="button"
            className="layers-section__add-button"
            onClick={this.handleAddButton}
          >
            <i className="fas fa-plus" />
          </button>
          <button
            type="button"
            className="layers-section__move-up"
            onClick={this.handleMoveUp}
          >
            <i className="fas fa-arrow-up" />
          </button>
          <button
            type="button"
            className="layers-section__move-down"
            onClick={this.handleMoveDown}
          >
            <i className="fas fa-arrow-down" />
          </button>
          <button
            type="button"
            className="layers-section__remove-button"
            onClick={this.handleRemoveButton}
          >
            <i className="fas fa-trash-alt" />
          </button>
        </div>
        <button
          type="button"
          className="layers-section__merge-button"
          onClick={this.handleMergeButton}
        >
          merge
        </button>
        <div className="layers-section__wrapper">
          {layerKeys.map((item, index) => (
            <button
              className={`layers-section__button ${
                index === activeLayer ? 'layers-section__button_active' : ''
              }`}
              id={`layer${layerKeys[index]}`}
              data-id={layerKeys[index]}
              type="button"
              key={layerKeys[index]}
              onClick={this.makeActive}
            >
              Layer {layerKeys[index] + 1}
            </button>
          ))}
        </div>
      </section>
    );
  }
}

Layers.propTypes = {
  updateLayerKeys: PropTypes.func.isRequired,
  updateActiveLayer: PropTypes.func.isRequired,
  framesData: PropTypes.instanceOf(Object).isRequired,
  framesArray: PropTypes.instanceOf(Array).isRequired,
  layersShortcuts: PropTypes.instanceOf(Object).isRequired,
  canvas: PropTypes.instanceOf(Object),
  layerKeys: PropTypes.instanceOf(Array).isRequired,
  activeLayer: PropTypes.number.isRequired,
  updateFramesData: PropTypes.func.isRequired,
};

Layers.defaultProps = {
  canvas: null,
};

const mapStateToProps = state => ({
  layersShortcuts: state.layers.layersShortcuts,
  framesArray: state.frames.framesArray,
  framesData: state.frames.framesData,
  canvas: state.canvas.canvasRef,
  layerKeys: state.layers.layerKeys,
  activeLayer: state.layers.activeLayer,
});

const mapDispatchToProps = (dispatch) => {
  const {
    updateLayerKeys,
    updateActiveLayer,
    updateFramesData,
  } = bindActionCreators(actions, dispatch);
  return {
    updateLayerKeys: (keys) => {
      updateLayerKeys(keys);
    },
    updateActiveLayer: (index) => {
      updateActiveLayer(index);
    },
    updateFramesData: (data) => {
      updateFramesData(data);
    },
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Layers);
