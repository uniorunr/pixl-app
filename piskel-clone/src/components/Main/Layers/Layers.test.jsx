import React from 'react';
import renderer from 'react-test-renderer';
import Layers from './Layers';

it('renders correctly', () => {
  const tree = renderer
    .create(
      <Layers
        updateLayerData={() => {}}
        layersShortcuts={{}}
        framesArray={[]}
        updateLayers={() => {}}
        framesData={{}}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
