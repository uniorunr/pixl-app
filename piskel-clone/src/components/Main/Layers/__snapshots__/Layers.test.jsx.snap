// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section
  className="layers-section"
>
  <div
    className="layers-section__controls"
  >
    <button
      className="layers-section__add-button"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fas fa-plus"
      />
    </button>
    <button
      className="layers-section__remove-button"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fas fa-trash-alt"
      />
    </button>
  </div>
  <div
    className="layers-section__wrapper"
  >
    <button
      className="layers-section__button layers-section__button_active"
      data-id={0}
      id="layer0"
      onClick={[Function]}
      type="button"
    >
      Layer 
      1
    </button>
  </div>
</section>
`;
